central/declarativeconfig/updater/auth_provider_updater.go

```diff
@@ -75,17 +76,30 @@ func (u *authProviderUpdater) DeleteResources(ctx context.Context, resourceIDsTo
 	var authProviderDeletionErr *multierror.Error
 	var authProviderIDs []string
 	for _, authProvider := range authProviders {
-		if err := u.authProviderRegistry.DeleteProvider(ctx, authProvider.GetId(), true, true); err != nil {
+		referencingGroups, err := u.groupDS.GetFiltered(ctx, func(group *storage.Group) bool {
+			return group.GetProps().GetAuthProviderId() == authProvider.GetId()
+		})
+		if err != nil || len(referencingGroups) > 0 {
+			if err == nil {
+				err = errox.ReferencedByAnotherObject.Newf("auth provider is still referenced by groups")
+			}
 			authProviderDeletionErr = multierror.Append(authProviderDeletionErr, err)
 			authProviderIDs = append(authProviderIDs, authProvider.GetId())
 
 			u.reporter.UpdateIntegrationHealthAsync(utils.IntegrationHealthForProtoMessage(authProvider, "", err,
 				u.idExtractor, u.nameExtractor))
+			authProvider.Traits.Origin = storage.Traits_DECLARATIVE_ORPHANED
+			if err = u.authProviderDS.UpdateAuthProvider(ctx, authProvider); err != nil {
+				authProviderDeletionErr = multierror.Append(authProviderDeletionErr, err)
+			}
```

Comment:
Hm, so we'd currently mark an auth provider as `orphaned` if a) the filtering for groups call fails unexpectedly b) it's referenced by either declarative groups or non-declarative groups.

I'd say we shouldn't set this to orphaned if `errors.Is(err, errox.ReferencedByAnotherObject) == false`.

On the other part with the references, I'd agree on the declarative groups, but I'm not too sure on the non-declarative ones: the resolution would be to first go within the UI, delete all of the groups created, and then deletion would succeed.

Do we want to make the distinction within declarative config for imperative groups, when in reality if you'd call the API to delete auth providers the associated groups would be simply deleted?

Maybe my perception at the moment is too harsh, and in reality `DeleteResource` will be called more often than we'd initially expect (e.g. people do really move resources around a lot), but my expectation would be that delete would be something done more explicit and the consequences (in this case deleting the imperative groups) would be understood and acceptable.

Reply:
Hmm, thinking about how we structure auth provider + groups declarative config:
* they are always defined together
* it's possible to delete group from auth provider
* not possible to delete auth provider without deleting the group

I start to think we should delete all(not only declarative) groups belonging to auth provider if auth provider is deleted. WDYT?